### ISSUE TITLE
[Arm64] JIT decrement compare pattern needs to be optimized

### ISSUE THREAD

**Issue by sdmaclea** [dotnet/coreclr#13784]
@dotnet/arm64-contrib @dotnet/jit-contrib 

While compiling buffer.cs
```
            n--;
            if (n != 0) goto MCPY06;
```

This code was generated
```
sub     x0, x0, #1
cmp     x0, #0
bne     G_M55607_IG21
```
should be

```
subs    x0, x0, #1
bne     G_M55607_IG21
```

**Comment by mikedn** [comment]
x64 does something like this in `Lowering::ContainCheckCompare`. ARM64 could do that too provided that CodeGen obeys the `GTF_SET_FLAGS` flag and generates `subs` instead of `sub`.

**Comment by sdmaclea** [comment]
@jashook Please assign this to me

**Comment by mikedn** [comment]
@sdmaclea FYI I'm considering changing a bit the implementation of this optimization so that most of the code is in `LowerCompare` and shared between ARM and XARCH. See #14027 (second commit)

**Comment by sdmaclea** [comment]
I have a draft of this optimization already.  I am trying to figure out why the original motivation is not supported.

It does seem like this belong in LowerCompare and not contain check.

**Comment by mikedn** [comment]
> I am trying to figure out why the original motivation is not supported.

Hmm?

> It does seem like this belong in LowerCompare and not contain check.

Yes, my opinion as well. It's a long story, this code was changed a few times and the flags thing is IMO still a bit of a hack.

**Comment by sdmaclea** [comment]
> Hmm?

```
n--;
if(n!=0) goto
```

The substitution is not happening because the `n != 0` looks like `GT_NE(LCL_VAR, 0)`.  Presumably this is a register allocator issue.  Since `n` is accessed in the only basic block within a loop, I would have assumed it would be promoted to a register first.

**Comment by sdmaclea** [comment]
> IMO still a bit of a hack.  

Looks like the XARCH hack is much cleaner than the ARM hack.  I was pleased how simple the change was for ARM64.

I wonder whether LSRA should be allocating the condition register.  This would affect the decision whether the Flags are in advance or delayed and set by the JUMP condition op (GT_NE et al.), or not at all (`cbz`, `cbnz`, `tbz`, `tbnz`).

**Comment by mikedn** [comment]
> I wonder whether LSRA should be allocating the condition register

AFAIR there were some discussions about handling the condition/flags register in the past. Not sure what the exact outcome was but it was in the "let's keep it simple" category. And at least on XARCH it would be rather strange to involve LSRA into this, there's a single flags registers and spilling/reloading it is not a reasonable option. So it's ultimately up to lowering (and even phases before it) to ensure that things are set up the proper way (e.g. there's no instruction between relop/jtrue that changes the flags).

What I don't like about the current implementation is that it doesn't actually remove the useless relop from IR, it leaves it there (and so it passes through LSRA) and relies on codegen to not generate a CMP instruction for it.

**Comment by mikedn** [comment]
> cbz, cbnz, tbz, tbnz

Eh, I was wondering about those recently. It seems to me that currently this cannot be represented accurately in IR, tbz/tbnz in particular.

**Comment by pgavlin** [comment]
> AFAIR there were some discussions about handling the condition/flags register in the past. Not sure what the exact outcome was but it was in the "let's keep it simple" category.

FWIW, we now model "writes flags"/"reads flags" using `GTF_{SET,USE}_FLAGS`, respectively. The more complete model would involve representing and calculating liveness for the same (probably using flags on `BasicBlock` to indicate that the flags are live-{in,out}) and taking these flags into account during side-effect analysis.

**Comment by sdmaclea** [comment]
Fixed by #14192
